### PR TITLE
Convert Black Github Action to Check instead of Force Merge

### PR DESCRIPTION
**Pull request by ew3361zh**
PR to change the Black Github Action to a check instead of running Black on the changed files and auto-merging in the updated code to the PR.

### Context
Several options were explored to solve the issue of the auto-merging black-formatting Github Action bypassing the other checks failing. This was allowing potentially linting/testing failing code to be merged into main. 

#### Current Github Actions Workflow
Github Actions (2)

Proposed Github Actions Workflow
Proposed Github Actions Flow

Issue(s) closed by this pull request: #1328

The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated: [ ]

### What
This PR updates the Github "Formatting and Linting" action to be a check. If it fails, it means black needs to be run locally on the code before it is pushed to Github. The requirements.txt file was also updated to make the version of black static at the current 24.2.0. This will help prevent unforeseen variations in the choices black is making in our Github action vs locally.

### Why
This will help prevent potentially testing/linting-failing code from being merged into main.

### How
The Github action check as written in this PR will operate similar to the flake8 linting action and will prevent merging to main if the code is not properly formatted.

### Test plan
I made the changes to the script and then intentionally changed some code in the `errors.py` file that violated the black guidelines and pushed that code to Github. I also did this with and without the `--diff` flag to see the level of detail it provides for why the code is failing the black guidelines. Finally, I ran black locally on my code and then pushed that to make sure it would fix the issues in agreement with how the github action would expect.

### Notes
[X] updated the versions of the actions/checkout and actions/setup-python to the v4s. They did not affect the other github actions and it was recommended for bug fixes and improvements.

### CODE REVIEW

File: README.md
Comment (Isgotkowitz) [inline README.md:6]:
```suggestion
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
```

**Review by Isgotkowitz — APPROVED** [review]
Looks great Niko. Added a suggestion for putting the Black formatting badge on the README.

When you tested out running unformatted code, did you find that the failed Action prevented PRs from being merged?

**Comment by ew3361zh** [comment]
> Looks great Niko. Added a suggestion for putting the Black formatting badge on the README.

Thanks, added!

> When you tested out running unformatted code, did you find that the failed Action prevented PRs from being merged?

It appeared like a normal failing action that would prevent merging but it's tough to tell without the approvals in place. Once I've got 2 approvals, I'll make it fail again and see what happens.

**Review by ltp23 — APPROVED** [review]
I haven't tried this out yet because I don't have access to this feature, but we could create a [protection rule](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/managing-a-branch-protection-rule) for the main branch. This would make sure that all workflows should pass before merging.

Alternatively, a workaround is to have only one workflow file and make all the exisiting workflows into jobs and order them in a certain way.

**Comment by ew3361zh** [comment]
Being addressed in PR #1500.

**Pull request closed without merging**
